this commit allows to "hijack" the Serenity reporter and on top of it…

eitzenbe opened the pull request:
… perform additional reporting/forwarding of test run progress to external tools (we use this in tiger to show a browser UI tracing the current testrun in realtime)

see https://github.com/gematik/app-Tiger

eitzenbe commented:
A screenie of the workflow UI can be found here: https://gematik.github.io/app-Tiger/Tiger-User-Manual.html#_workflow_ui

wakaleo commented:
Interesting. Could you add some unit or smoke tests to demonstrate how it works?

eitzenbe commented:
Sure, I see no tests at all for the SerenityReporter so far, but will get som ebasic stuff up'n running to show and verify the feature request ;)

eitzenbe commented:
@wakaleo I have refactored the code and removed the callback feature but refactored the runner to have a base runner that is extendable and contains most code and also made the Reporter's major functionality reachable from an extending class. This doesn't change any functionality, so no need for unit tests there. I am attaching two files to illustrate how this can be used to extend the reporter plugin .... Any chance this could go in with the next release? 3.6.x
[TigerSerenityReportPlugin.java.txt](https://github.com/serenity-bdd/serenity-core/files/10560620/TigerSerenityReportPlugin.java.txt)
[TigerCucumberListener.java.txt](https://github.com/serenity-bdd/serenity-core/files/10560621/TigerCucumberListener.java.txt)

wakaleo commented:
@cliviu could you take a look at this?

eitzenbe commented:
@cliviu any timeline for review? We would need this for Tiger release and customers are waiting ;)

cliviu commented on serenity-cucumber/src/main/java/io/cucumber/junit/CucumberSerenityBaseRunner.java:
```diff
@@ -0,0 +1,346 @@
+import net.thucydides.core.environment.SystemEnvironmentVariables;
+import net.thucydides.core.steps.StepEventBus;
+import net.thucydides.core.util.EnvironmentVariables;
+import org.jetbrains.annotations.NotNull;
```
why the jetbrains @NotNull annotation?

eitzenbe replied:
my bad will be removed

eitzenbe commented on serenity-cucumber/src/main/java/io/cucumber/junit/CucumberSerenityBaseRunner.java:
```diff
@@ -0,0 +1,346 @@
+public class CucumberSerenityBaseRunner extends ParentRunner<ParentRunner<?>> {
+
+
+    protected EventBus bus;
```
Would you mind to elaborate? in what aspect is package friendly a better approach here?

cliviu replied:
not package friendly but private

eitzenbe replied:
But we need access to these fields if we want to customize the runner, else the base class has no use at all...

eitzenbe replied:
I could try to minimize the accessability and make the fields private, getters protected and maybe I can circumvent having any PROTECTED setter method  (plugins and context fields are troublesome)

cliviu replied:
yes, this is what i mean. even if is arguable, i still find the protected fields a bad smell in code

eitzenbe replied:
done see commit below ;)

eitzenbe commented:
@cliviu new push, should resolve your remarks ;)

cliviu commented on serenity-cucumber/src/main/java/io/cucumber/junit/CucumberSerenityBaseRunner.java:
```diff
@@ -0,0 +1,376 @@
+    private List<Feature> features;
+    private CucumberExecutionContext context;
+
+    protected boolean multiThreadingAssumed = false;
```
still protected

eitzenbe replied:
I think it was protected in the original code, svejedno I have changed it to private see new push

cliviu commented on serenity-cucumber/src/main/java/io/cucumber/junit/CucumberSerenityBaseRunner.java:
```diff
@@ -0,0 +1,376 @@
+    protected boolean multiThreadingAssumed = false;
+
+
+    static ThreadLocal<RuntimeOptions> RUNTIME_OPTIONS = new ThreadLocal<>(); // NOSONAR
```
looks like this comment is not needed here

eitzenbe replied:
This is needed as Sonar code analysis is biting us here with a false positive warning

cliviu commented on serenity-cucumber/src/main/java/io/cucumber/junit/CucumberSerenityBaseRunner.java:
```diff
@@ -0,0 +1,376 @@
+
+    static ThreadLocal<RuntimeOptions> RUNTIME_OPTIONS = new ThreadLocal<>(); // NOSONAR
+
+    private static RuntimeOptions DEFAULT_RUNTIME_OPTIONS; // NOSONAR
```
same here

eitzenbe commented:
@cliviu new push, should resolve also your second set of remarks ;)
